Core/WebCacheManager.swift

```diff
@@ -33,14 +33,38 @@ public class WebCacheManager {
         return WKWebsiteDataStore.default()
     }
 
-    public static func consumeCookies() {
-        guard #available(iOS 11, *) else { return }
+    public static func consumeCookies(completion: @escaping () -> Void) {
+        guard #available(iOS 11, *) else {
+            completion()
+            return
+        }
 
         let cookieStorage = CookieStorage()
-        for cookie in cookieStorage.cookies {
-            WebCacheManager.dataStore.httpCookieStore.setCookie(cookie)
+        let cookies = cookieStorage.cookies
+        
+        guard !cookies.isEmpty else {
+            completion()
+            return
+        }
+        
+        let semaphore = DispatchSemaphore(value: 0)
+                        
+        for cookie in cookies {
+            WebCacheManager.dataStore.httpCookieStore.setCookie(cookie) {
+                semaphore.signal()
+            }
+        }
+        
+        DispatchQueue.global(qos: .userInitiated).async {
+            for _ in 0 ..< cookies.count {
+                semaphore.wait()
+            }
+            
+            DispatchQueue.main.async {
+                cookieStorage.clear()
+                completion()
+            }
```

Comment:
Minor nitpick / suggestion:

When waiting for multiple items, I find that using DispatchGroup makes code a little more readable. Something like:

``` Swift
        let group = DispatchGroup()

        for cookie in cookies {
            group.enter()
            WebCacheManager.dataStore.httpCookieStore.setCookie(cookie) {
                group.leave()
            }
        }

        group.notify(queue: DispatchQueue.main) {
            cookieStorage.clear()
            completion()
        }
```

Reply:
Yes, this is very good suggestion. Since this is already merged I'm not going to touch that right now, but I'll track it internally and address it next time I'm doing some cleaning.